ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,596 @@
+
+newtype NameMK k v = NameMK String
+
+-- | A general interface to mapkinds.
+--
+-- In some cases where @mk@ is not specialised to a concrete mapkind like
+-- @'ValuesMK'@, there are often still a number of operations that we can
+-- perform for this @mk@ that make sense regardless of the concrete mapkind. For
+-- example, we should always be able to map over a mapkind to change the type of
+-- values that it contains. This class is an interface to mapkinds that provides
+-- such common functions.
+type IsMapKind :: MapKind -> Constraint
+class ( forall k v. (Eq       k, Eq       v) => Eq       (mk k v)
+      , forall k v. (NoThunks k, NoThunks v) => NoThunks (mk k v)
+      , forall k v. (Show     k, Show     v) => Show     (mk k v)
+      ) => IsMapKind mk where
+  emptyMK :: forall k v. (Ord k, Eq v) => mk k v
+  default emptyMK :: forall k v. Monoid (mk k v) => mk k v
+  emptyMK = mempty
+
+  mapMK :: forall k v v'. (Ord k, Eq v, Eq v') => (v -> v') -> mk k v -> mk k v'
+  default mapMK :: forall k v v'. (Functor (mk k)) => (v -> v') -> mk k v -> mk k v'
+  mapMK = fmap
+
+instance IsMapKind Canonical where
+  emptyMK = Canonical
+  mapMK _ Canonical = Canonical
+
+instance IsMapKind EmptyMK where
+  emptyMK = EmptyMK
+  mapMK _ EmptyMK = EmptyMK
+
+instance IsMapKind KeysMK where
+  emptyMK = KeysMK mempty
+  mapMK _ (KeysMK ks) = KeysMK ks
+
+instance IsMapKind ValuesMK where
+  emptyMK = ValuesMK mempty
+  mapMK f (ValuesMK vs) = ValuesMK $ fmap f vs
+
+instance IsMapKind TrackingMK where
+  emptyMK = TrackingMK mempty mempty
+  mapMK f (TrackingMK vs d) = TrackingMK (fmap f vs) (fmap f d)
+
+instance IsMapKind DiffMK where
+  emptyMK = DiffMK mempty
+  mapMK f (DiffMK d) = DiffMK $ fmap f d
+
+instance IsMapKind SeqDiffMK where
+  emptyMK = SeqDiffMK empty
+  mapMK f (SeqDiffMK ds) = SeqDiffMK $ mapDiffSeq f ds
+
+instance Ord k => Semigroup (KeysMK k v) where
+  KeysMK l <> KeysMK r = KeysMK (l <> r)
+
+instance Ord k => Monoid (KeysMK k v) where
+  mempty = KeysMK mempty
+
+instance Functor (DiffMK k) where
+  fmap f (DiffMK d) = DiffMK $ fmap f d
+
+{-------------------------------------------------------------------------------
+  Serialization Codecs
+-------------------------------------------------------------------------------}
+
+-- | This class provides a 'CodecMK' that can be used to encode/decode keys and
+-- values on @'LedgerTables' l mk@
+type CanSerializeLedgerTables :: LedgerStateKind -> Constraint
+class CanSerializeLedgerTables l where
+  codecLedgerTables :: LedgerTables l CodecMK
+  default codecLedgerTables :: LedgerTablesAreTrivial l => LedgerTables l CodecMK
+  codecLedgerTables = trivialLedgerTables
+
+-- | Default encoder of @'LedgerTables' l ''ValuesMK'@ to be used by the
+-- in-memory backing store.
+valuesMKEncoder ::
+     ( HasLedgerTables l
+     , CanSerializeLedgerTables l
+     )
+  => LedgerTables l ValuesMK
+  -> CBOR.Encoding
+valuesMKEncoder tables =
```

Comment:
Are `valuesMKEncoder` and `valuesMKDecoder` going to be tested in later PRs?

Reply:
Yeah, the tests come in a future PR